Tickets/dm 36310 (#3)
lsst-dm/dmtn-168: index.rst

Where $projectID is *panda-dev-1a74*.  Then it is passed to the container on the POD nodes via the secret name
*gcs-access*, with the environmental variable **GOOGLE_APPLICATION_CREDENTIAL** pointing to the json file.

Real-time Logging for Pilot Jobs
--------------------------------
Pilot logs associated with payload jobs are written onto the GCS bucket **drp-us-central1-logging**. 
We also provide (near)real-time logging for pilot logs on Google Cloud Logging, regardless whether 
the pilot job is idling or running some payload jobs. It could help provide real-time 
logging of whole pilot jobs as well as the parental pilot wrapper jobs. In addition, it is very useful in case pilot logs 
could not be written onto the GCS for some reasons such as pilot jobs being killed or problems with GCS.

The real-time logging for pilot jobs is implemented in the pilot starter python script *pilot3_starter.py*, 

> with the class *RealTimeLogger*. The pilot wrapper log (which already icnludes the pilot log itself) file */tmp/wrapper-wid.log*, 

Reviewer: ```suggestion
with the class *RealTimeLogger*. The pilot wrapper log (which already includes the pilot log itself) file */tmp/wrapper-wid.log*, 
```

is parsed and filtered, then written into the log name **Panda-WorkerLog**, different from the log name for payload job logs.

Usually only logs containg a valid time stamp at the beginning is kept and written onto Google Cloud Logging.
For each log, the hostname is added. And PanDA job ID and job name are also added if available.

Normally the log severity level is just copied from the pilot log level. However, if the payload stderr dump message 

> indicates that the payload failed, the log severity level would be set to ERROR for that msg and also pilot logs 

Reviewer: ```suggestion
indicates that the payload failed, the log severity level would be set to ERROR for that message and also pilot logs 
```

containing either *"| add_error_codes"* or *"| perform_initial_payload_error_analysis"*.

For the logs of payload stderr dump, the last 10 non-blank lines (though not containing a time stamp) 

> following the line of *"| add_error_codes"* are also added and appended into the line of *"| add_error_codes"*.

Reviewer: ```suggestion
following the line of *"| payload stderr dump"* are also added and appended into the line of *"| payload stderr dump"*.
```

Author: There should be 3 types of logs whose severity level might be changed to ERROR in case of payload failure. I missed one type containing **"| payload stderr dump"**. I just fixed it.

under the job status such as *running*, *finished*, or *failed*, will show the list of jobs in that status. You can
check each job details by following *the PanDA ID number*.

Real-time Logging for Payload Jobs
----------------------------------
The Rubin jobs on the PanDA queues are also provided with (near)real-time logging on Google Cloud Logging 
in log name **Panda-RubinLog**.
Once the jobs have been running on the PandDA queues, users can check the json format job logs on 
`the Google Logs Explorer <https://console.cloud.google.com/logs>`_.

The enviromental variable **REALTIME_LOGFILES** defines the json filename to be produced by Rubin payload jobs.
A pilot job continuously reads this json file *every 5 seconds*, and sends the found lines together with associated 
PanDA job ID and task ID.

The envvar **REALTIME_LOGFILES=payload-log.json** is set in the Kubernetes job yaml file for each queue, 

> undet the *container env* section, as well as 3 other real-time envvars:

Reviewer: ```suggestion
under the *container env* section, as well as 3 other real-time envvars:
```

> - *USE_REALTIME_LOGGING=yes*

Reviewer: ```suggestion
- *USE_REALTIME_LOGGING="yes"*
```

> - *REALTIME_LOGGING_SERVER=google-cloud-logging*

Reviewer: ```suggestion
- *REALTIME_LOGGING_SERVER="google-cloud-logging"*
```

> - *REALTIME_LOGNAME=REALTIME_LOGNAME*

Reviewer: ```suggestion
- *REALTIME_LOGNAME="Panda-RubinLog"*
```

which correspond to the following 3 options in the main pilot python script *pilot.py*:

> - *--use-realtime-logging*

Reviewer: ```suggestion
- ``--use-realtime-logging``
```

> - *--realtime-logging-server*

Reviewer: ```suggestion
- ``--realtime-logging-server``
```

> - *--realtime-logname*

Reviewer: ```suggestion
- ``--realtime-logname``  
```
